metric_consolidator.go

```diff
@@ -0,0 +1,98 @@
+package gostatsd
+
+import (
+	"context"
+	"time"
+
+	"github.com/tilinna/clock"
+)
+
+// MetricConsolidator will consolidate metrics randomly in to a slice of MetricMaps, and send the slice to the provided
+// channel.  Run can be started in a long running goroutine to perform flushing, or Flush can be called externally.
+type MetricConsolidator struct {
+	maps          chan *MetricMap
+	sink          chan<- []*MetricMap
+	flushInterval time.Duration
+}
+
+func NewMetricConsolidator(spots int, flushInterval time.Duration, sink chan<- []*MetricMap) *MetricConsolidator {
+	mc := &MetricConsolidator{}
+	mc.maps = make(chan *MetricMap, spots)
+	for i := 0; i < spots; i++ {
+		mc.maps <- NewMetricMap()
+	}
+	mc.flushInterval = flushInterval
+	mc.sink = sink
+	return mc
+}
+
+func (mc *MetricConsolidator) Run(ctx context.Context) {
+	t := clock.NewTicker(ctx, mc.flushInterval)
+	defer t.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-t.C:
+			mc.Flush(ctx)
+		}
+	}
+}
+
+func (mc *MetricConsolidator) Flush(ctx context.Context) {
+	var mms []*MetricMap
+	for i := 0; i < cap(mc.maps); i++ {
+		mm := mc.getMap(ctx)
+		if mm == nil {
+			// Put everything back, so we're consistent, just in case.  No need to check for termination,
+			// because we know it will fit exactly.
+			for _, mm := range mms {
+				mc.maps <- mm
+			}
+			return
+		}
+		mms = append(mms, mm)
+	}
+
+	// Send the collected data to the sink before putting new maps in place.  This allows back-pressure
+	// to propagate through the system, if the sink can't keep up.
+	select {
+	case mc.sink <- mms:
+	case <-ctx.Done():
```

Comment:
In theory we should be putting the original maps back because it's "correct".  In practice, it's shutdown time and data quality doesn't matter at that point.  I didn't bother handling each case explicitly because it's simpler.